adhocteam/Head-Start-TTADP
# Cm 102 fix objectives delete error

**Description of change**
Fixes bug where objectives from saved goals were being provisioned in a new report when the saved goal was added. If you tried to delete these goals and then save the report, it caused a 500 response because it would try to delete an objective that was being used by another report.

Thanks @rahearn for figuring out how to reproduce the bug and @jasalisbury for pinning down where exactly the fix should go.

The solution had 2 parts.
1. Change the `removeUnusedObjectivesGoalsFromReport` function in `src/services/goals.js` to not delete objectives that were used by other reports.
2. Update the `ActivityReports` model so the goals virtual field only returns objectives that belong to that report instead of all objectives associated with a report.

**How to test**
Checkout the main branch and complete a activity report, submit it, and approve it.
Create a second activity report using the same grantee from the first report.
Add the goal that was saved in the first report. Save the report and refresh the page.
You will see objectives from the first and second report.
Save the draft again.
At this point, stop  the app, checkout this branch and start the app again.
Refresh the page in the browser and you will still see both objectives.
Delete the objective from the completed report and save the report.
It should save successfully.
Create a third report still running on this branch. Use the same grantee and add the goal from the first report.
Refresh the page. You should only see the objective you just added, not both of them.

**Issue(s)**
* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-102

**Checklist**

    - API methods
    - Boundary and Data Flow Diagrams
    - [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions with the [Nygard template](https://github.com/joelparkerhenderson/architecture_decision_record/blob/master/adr_template_by_michael_nygard.md)
    - OSCAL templates completed when security controls are implemented or modified
